routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,185 @@
+#include "routing/bicycle_directions.hpp"
+#include "routing/car_model.hpp"
+#include "routing/router_delegate.hpp"
+#include "routing/routing_result_graph.hpp"
+#include "routing/turns_generator.hpp"
+
+#include "geometry/point2d.hpp"
+
+#include "indexer/ftypes_matcher.hpp"
+#include "indexer/index.hpp"
+#include "indexer/scales.hpp"
+
+namespace
+{
+using namespace routing;
+using namespace routing::turns;
+
+class AStarRoutingResultGraph : public IRoutingResultGraph
+{
+public:
+  AStarRoutingResultGraph(IRoadGraph::TEdgeVector const & routeEdges,
+                          AdjacentEdgesMap const & adjacentEdges,
+                          TUnpackedPathSegments const & pathSegments)
+    : m_routeEdges(routeEdges), m_adjacentEdges(adjacentEdges), m_pathSegments(pathSegments),
+      m_routeLength(0)
+  {
+    for (auto const & edge : routeEdges)
+    {
+      m_routeLength += MercatorBounds::DistanceOnEarth(edge.GetStartJunction().GetPoint(),
+                                                       edge.GetEndJunction().GetPoint());
+    }
+  }
+
+  ~AStarRoutingResultGraph() {}
+
+  virtual TUnpackedPathSegments const & GetSegments() const override
+  {
+    return m_pathSegments;
+  }
+
+  virtual void GetPossibleTurns(TNodeId node, m2::PointD const & ingoingPoint,
+                                m2::PointD const & junctionPoint,
+                                size_t & ingoingCount,
+                                TTurnCandidates & outgoingTurns) const override
+  {
+    if (node >= m_routeEdges.size() || node >= m_adjacentEdges.size())
+    {
+      ASSERT(false, ());
```

Comment:
What the point to compare node with m_adjacentEdges.size()? adjacentEdges.size() is a total number of nodes in a hashmap, whereas node may exists in this hashmap, even if it's greater that the adjacentEdges.size(), right?